af/fastpath/fastpath/db.py

```diff
@@ -334,3 +228,94 @@ def tf(v: bool) -> str:
     #         is_ssl_expected = "0"
     # else:
     #     is_ssl_expected = "2"
+
+
+@metrics.timer("clickhouse_upsert_openvpn_obs")
+def clickhouse_upsert_openvpn_obs(
+    msm,
+    scores,
+    measurement_uid: str,
+) -> None:
+    sql_insert = dedent(
+        """\
+    INSERT INTO obs_openvpn (
+    anomaly,
+    bootstrap_time,
+    confirmed,
+    error,
+    failure,
+    input,
+    measurement_start_time,
+    measurement_uid,
+    obfuscation,
+    platform,
+    probe_asn,
+    probe_cc,
+    probe_network_name,
+    provider,
+    remote,
+    report_id,
+    resolver_asn,
+    resolver_ip,
+    resolver_network_name,
+    software_name,
+    software_version,
+    success,
+    tcp_connect_status_success,
+    test_runtime,
+    test_start_time,
+    transport
+    ) VALUES
+        """
+    )
+
+    def nn(d: dict, k: str) -> str:
+        """Get string value and never return None"""
+        v = d.get(k)
+        return "" if v is None else v
+
+    def tf(v: bool) -> str:
+        return "t" if v else "f"
+
+    asn = int(msm["probe_asn"][2:])  # AS123
+    measurement_start_time = datetime.strptime(
+        msm["measurement_start_time"], "%Y-%m-%d %H:%M:%S"
+    )
+    test_start_time = datetime.strptime(msm["test_start_time"], "%Y-%m-%d %H:%M:%S")
+    tk = dget_or(msm, "test_keys", {})
+
+    anomaly = nn(msm, "success") == True
+    tcp_connect_status_success = "t" #FIXME
+    row = [
+        anomaly,
+        dget_or(tk, "bootstrap_time", 0),
+        "f",
+        nn(msm, "error"),
+        nn(msm, "failure"),
+        nn(msm, "input"),
+        measurement_start_time,
+        measurement_uid,
+        nn(tk, "obfuscation"),
+        nn(msm, "platform"),
+        asn,
+        nn(msm, "probe_cc"),
+        nn(msm, "probe_network_name"),
+        nn(tk, "provider"),
+        nn(tk, "remote"),
+        nn(msm, "report_id"),
+        nn(msm, "resolver_asn"),
+        nn(msm, "resolver_ip"),
+        nn(msm, "resolver_network_name"),
+        nn(msm, "software_name"),
+        nn(msm, "software_version"),
+        nn(msm, "success"),
+        tcp_connect_status_success,
+        dget_or(msm, "test_runtime", 0),
+        nn(tk, "transport"),
+    ]
```

Comment:
Using this pattern of inserting using a list makes me a little bit nervous. If we add or remove columns we need to be sure that the order of the fields in the row is preserved. It's also very hard to visually validate if the order is correct even currently.

I would suggest using instead the dict based insertion so that you ensure you are putting the values inside of the correct column.

Reply:
Updated